### ISSUE TITLE
Add a function to understand netcdf times

### ISSUE THREAD

**Issue by adrfantini** [ropensci/tidync#54]
A function such the ones proposed [here](https://stackoverflow.com/questions/46001573/convert-a-netcdf-time-variable-to-an-r-date-object/46011663#46011663) might be a useful addition to the packages. These help translating NetCDF times into R timedates.
Every time, getting time data from a NetCDF is a pain... one might want to look at how NCL, GrADS and the likes do it, but a simple approach like that in the link above might suffice.

**Comment by mdsumner** [comment]
This is a neglected area for me, I appreciate you asking and pointing out that thread - the thing that scares me is  non-standard calendars and supporting those properly. The "tidync way" to do this is to put the unit-aware version of time on the axis (transform) table for a time dimension, so that users can write expressions like `hyper_filter(time = unit_time > as.Date("2001-01-01"))`. It's trivial to do that for a "days since <epoch>" metadata item converted to Date or POSIXct, but I am worried about presenting an interpretation that's actually not correct in certain contexts. I would **always** have `index`, `<name-of-var>` and then `unit_<name-of-var>` or something similar to ensure that the raw index, the raw value, and then the unit version were available - but I'm terrified of getting the unit-ed version wrong. The RNetCDF function `utcal.nc` seems to take this on whole-heartedly, and the climate science community do know the details. 

In short I'm very happy to do this but I'm lacking confidence and expertise, so any input you can have is appreciated!

**Comment by matteodefelice** [comment]
Yes, I would love that but I agree that taking into account *all* the possible cases is a big challenge. I often use the functions is `ncdf4.helpers` and it would be nice to try to add the possibility to add a meaningful unit to the time axis.

**Comment by adrfantini** [comment]
A function that covers *most* cases would still be useful tho, right? Look for example at CDO, it can correctly read the times from any NetCDF files and I have never seen it fail doing so. There must be a way that covers >95% of usecases, right?
Maybe including it in `ncdf4.helpers` would be the best idea, tho.

**Comment by mdsumner** [comment]
I'm happy to put this in,  I'm working towards it slowly  -  a key part will be that the unit version will be able to be re-interpreted if needed -  it's just another column on an axis table.  Any examples you can outline using ncdf4.helpers or RNetCDF or similar   would be helpful, especially if the standard POSIXct or Date conversion would not be the correct one.

**Comment by mdsumner** [comment]
Oops didn't mean to close...

**Comment by adrfantini** [comment]
I'm not sure I understand how I could help. Isn't the `nc.get.time.series` function from `ncdf4.helpers` a good base to start from? What's the problem with it?
You can find it here:
https://github.com/cran/ncdf4.helpers/blob/master/R/helpers.r

**Comment by mdsumner** [comment]
Outlining examples is helpful. It provides a basis for knowing the functionality is on target and relevant. (I didn't mention any problems with that function, I haven't explored it at all until now. It looks like it's very useful, something that should be in RNetCDF or ncdf4 directly - do you think it covers all cases?).

**Comment by adrfantini** [comment]
The approach `stars` uses is to use package `units`. This works well, but AFAIK (I asked, waiting for answer) does not support the weird calendars some of us love.

A function like this is very general, if the calendar is gregorian:

> getNcTime <- function(nc) { 
>     require(units)
>     require(ncdf4)
>     options(warn=1) #show warnings by default
>     if (is.character(nc)) nc <- nc_open(nc)
>     ncdims <- names(nc$dim) #get netcdf dimensions
>     timevar <- ncdims[which(ncdims %in% c("time", "Time", "datetime", "Datetime", "date", "Date"))] #find (first) time variable
>     if (length(timevar) > 1) {
>         warning(paste("Found more than one time var. Using the first:", timevar[1]))
>         timevar <- timevar[1]
>     }
>     if (length(timevar)!=1) stop("ERROR! Could not identify the correct time variable")
>     times <- ncvar_get(nc, timevar) #get time data
>     timeatt <- ncatt_get(nc, timevar) #get attributes
>     timeunit <- timeatt$units
>     units(times) <- make_unit(timeunit)
>     as.POSIXct(time)
> }

**Comment by mdsumner** [comment]
I'm afraid the general nature of the landscape means it's fraught with danger. I want it to be easy with tidync to apply your own interpretation, but I want to avoid applying them automatically. This packages doesn't know anything about spatial coordinates or other units, so for now I want to keep things raw.  If there comes general support that will always work, or fail hard, I'll pursue it - but I just don't think it's possible. I see tidync being used to write (or learn to write) other tools that are for specific and known issues with types of data sources. 

I appreciate you providing this, but I'll need to see if it's going to be ok to use. I think if units gets really sophisticated we could leverage it for NetCDF, but I don't believe it's possible to ever do it without user checking/intervention, given the vagaries of files and assumptions I've seen.

**Comment by adrfantini** [comment]
I totally understand the approach `tidync` wants to pursue, don't worry about that. Feel free to close if you do not want to pursue this further. That's why software people either love or hate NetCDF - it's so versatile you can do everything with it, yet coding for it is very hard, despite conventions.

**Comment by mdsumner** [comment]
Cool, I'll keep it open -  it's basically unsolved and I'd appreciate any more discussion that might come along ;)

**Comment by dlebauer** [comment]
+1 to the idea. I regularly have to add a few lines in my code to convert indices to dates. I'll try out some of the existing functions, though would it reduce the scope of the challenge to limit support to CF-compliant time dimensions (http://cfconventions.org/Data/cf-conventions/cf-conventions-1.7/build/ch04s04.html)?

**Comment by adrfantini** [comment]
In my experience `stars::read_stars()` understands time well, at least for my NetCDF files.

In addition to the previous function, I sometimes find myself resorting to the one below, since it can handle 36_day calendars:

```
getNcTime2 <- function(nc) {
    require(PCICt)
    require(ncdf4)
    options(warn=1) #show warnings by default
    if (is.character(nc)) nc <- nc_open(nc)
    ncdims <- names(nc$dim) #get netcdf dimensions
    timevar <- ncdims[which(ncdims %in% c("time", "Time", "datetime", "Datetime", "date", "Date"))] #find (first) time variable
    if (length(timevar) > 1) {
        warning(paste("Found more than one time var. Using the first:", timevar[1]))
        timevar <- timevar[1]
    }
    if (length(timevar)!=1) stop("ERROR! Could not identify the correct time variable")
    times <- ncvar_get(nc, timevar) #get time data
    timeatt <- ncatt_get(nc, timevar) #get attributes
    timedef <- strsplit(timeatt$units, " ")[[1]]
    timeunit <- timedef[1]
    tz <- timedef[5]
    cal <- timeatt$calendar
    timestart <- strsplit(timedef[4], ":")[[1]]
    if (length(timestart) != 3 || timestart[1] > 24 || timestart[2] > 60 || timestart[3] > 60 || any(timestart < 0)) {
        warning("Warning:", timestart, "not a valid start time. Assuming 00:00:00")
        timedef[4] <- "00:00:00"
    }
    if (! tz %in% OlsonNames()) {
        warning("Warning:", tz, "not a valid timezone. Assuming UTC")
        tz <- "UTC"
    }
    timestart <- as.PCICt(paste(timedef[3], timedef[4]), tz=tz, cal=cal)
    f <- switch(tolower(timeunit), #Find the correct lubridate time function based on the unit
        seconds=1, second=1, sec=1,
        minutes=60, minute=60, min=60,
        hours=3600,     hour=3600,     h=3600,
        days=86400,       day=86400,       d=86400,
        NA
    )
    suppressWarnings(if (is.na(f)) stop("Could not understand the time unit format"))

    as.POSIXct(timestart + times*f)
}
```

**Comment by adrfantini** [comment]
@dlebauer Limiting to CF-c time units can help, but consider for example the case of different calendars. `R` has very limited support for the weird calendars that the convention supports. It is quite a challenge.

**Comment by mdsumner** [comment]
I don't think I want this in tidync, there's no automated unit-support here and I think that's how it should be.  tidync could better expose the underlying metadata tags, but that's a different topic

**Comment by pvanlaake** [comment]
If there is any interest to reopen this issue, the recently released package CFtime can handle all CF Metadata Conventions calendars, including units "hours", "minutes" and "seconds" in addition to the more common "days". It is pretty complete, it easily hits @adrfantini 's 95% of use cases threshold. Operation is as easy as:

```
nc <- nc_open("some.nc")
cf <- CFtime(nc$dim$time$units, nc$dim$time$calendar, nc$dim$time$vals)

# Vector of character timestamps that can be used for dimnames() etc
names <- CFtimestamp(cf, "timestamp")
```

That said, several of the CF calendars are not compatible with POSIXt so `as.Date()` will (gracefully) choke on a valid date like "2023-02-30" in the "360_day" calendar - there's no getting away from that.

**Comment by mdsumner** [comment]
sure, that looks good - I'll have a look and appreciate any leads

tidync is due a major reboot, and would be awesome to have it cover hdf5 and zarr and all that stuff

**Comment by pvanlaake** [comment]
I've been digging around in the code today and I think I have a way forward for "loose coupling".

Referring back to Michael's message of 2 Jan 2018 in this thread, I understand that the code is agnostic to dimension details by (current) design. Going for a description of the "time" dimension would break that design, so I would not argue for a solution in that direction. We do have to recognise, though, that NetCDF files are self-describing through the various attribute sets so all necessary details are present in the file; `tidync` is currently not providing access to those attributes and that could be added with a few lines of code (forgive my ignorance if I am wrong here). In `tidync.character()` a single line of code would provide access to all attributes:

```
tidync.character <- function(x, what, ...) {

(...)

out <- list(source = meta$source, 
              axis = meta$axis, 
              grid = meta$grid,
              dimension = meta$dimension, 
              variable = variable,
              attribute = meta$attribute)                # this line

(...)

}
```

In a "tighter" way of coupling, and after verifying that we indeed have a "time" dimension variable, there should be code snippets here and there that basically do this (code specifics will be off, this is transcribed console code):

```
units <- (x$attribute %>% filter(variable == "time", name == "units") %>% select("value"))[[1]]$units
calendar <- (x$attribute %>% filter(variable == "time", name == "calendar") %>% select("value"))[[1]]$calendar
offsets <- unlist(x %>% activate("time") %>% hyper_array())
cf <- CFtime(units, calendar, offsets)
times <- CFtimestamp(cf, "timestamp")
```

The `times` object can now be used in `dimnames()` on the data that is read by the `hyper*()` set of functions (before conversion to a `tibble`, if so requested).

A big advantage to this is that you can apply the same principle of attribute inspection to describe the other dimensions of the file (and indeed the variables). This requires importing the `CFtime` package into `tidync`.

The "loose coupling" I started off with would only expose the attributes through some mechanism such that users of `tidync` can apply `CFtime` workflow themselves. This would be the easiest to implement and with a few good examples in a vignette would make this quite accessible.

**Comment by pvanlaake** [comment]
This would be the function to get a loose coupling between `tidync` and `CFtime` such that `hyper_*()` objects can have a proper time indication:

```
time_dimnames <- function(tnc) {
  require(CFtime)

  # Get all attributes with predictable case
  attributes <- tnc$attribute |> mutate(low_name = str_to_lower(name), low_var = str_to_lower(variable)) 
  
  # Find a "time" dimension
  if (attributes |> filter(low_var == "time") |> count() > 0) {
    time <- attributes |> filter(low_var == "time") |> slice(1) |> pull(variable)
  } else {
    axes <- attributes |> filter(low_name == "axis")
    ndx <- which((axes |> select(value)) == "T")
    if (length(ndx) > 0) {
      time <- axes |> slice(ndx) |> pull(variable)
    } else {
      return(NULL)
    }
  }
  
  # Build CFtime object
  units <- unlist(attributes |> filter(variable == time, low_name == "units") |> pull(value))
  if (is.null(units)) return(NULL)
  calendar <- unlist(attributes |> filter(variable == time, low_name == "calendar") |> pull(value))
  if (is.null(calendar)) calendar <- "standard"         # COARDS does not insert attribute for standard calendar
  offsets <- (tnc |> activate(time) |> hyper_array())[[1]]
  cf <- CFtime(units, calendar, offsets)

  tibble(time = offsets, timestamp = CFtimestamp(cf, "timestamp"))
}
```

As input the function takes a tidync object (`tnc`) and it produces a tibble with the "time" values from the file and a corresponding character timestamp following the CF calendar of the data set. That output can be used to assign dimnames to the data produced by a `hyper_*()` function. The workflow for a tibble is then:

```
tnc <- tidync("some.nc")
tdn <- time_dimnames(tnc)
data <- (tnc |> activate("my_var") |> hyper_tibble())$my_var
if (!(is.null(tdn)) data <- data |> left_join(tdn)
```

For the other `hyper_*()` variants there should be some accommodation for any applied `hyper_filter()` but that should not be a big issue.

All of the above assumes that attributes are exposed through the tidync object, using the one-line hack I described before. Using `tnc$attribute` is of course not very desirable so there should be a proper function call to give access to that member of the class.

All of the above notwithstanding, it now appears to me that the better solution would be to integrate `CFtime` into `tidync` and produce proper time information right off the bat. That could be set up in such a way that the fallback option on failure is to produce the current output, e.g. the raw offsets. Happy to support if a decision to that effect is made.

**Comment by mdsumner** [comment]
this sounds good, realistically I won't be able to look into this for a few weeks but very happy to review a PR, and discuss

CFtime looks good to me, do you know if it's got a public repo? I can see the source at cran/CFtime but would like to know where it's hosted for dev 🙏

**Comment by pvanlaake** [comment]
CFtime lives here: https://github.com/pvanlaake/CFtime and I am the developer.

I have forked tidync and I'll start looking into building CFtime into tidync. It seems to me that the axis transforms are the place to start: if the timestamp can be placed in the time axis tibble then we should be almost there, right? Any suggestions would be appreciated.

I first want to put out a new release of CFtime before I get into this so I won't be bugging you until later this month.

**Comment by mdsumner** [comment]
oh gosh, I'm so sorry - I looked closely at the package and the names too, I didn't make the connection and couldn't find its home 🙏

I appreciate all the details and will follow up 👌

**Comment by pvanlaake** [comment]
Patch is committed in the fork mentioned above. `hyper_array()`, `hyper_tibble()` and `hyper_tbl_cube()` now all show real date/time information as character strings (but not necessarily POSIXt compliant). This is achieved through 3 basic steps:

1. Expose the attributes of the NetCDF file in `tidync()` (this also resolves issue #99 although there should be a formal public access method added).
2. Build the timestamp information in `hyper_transforms()` and attaching it to the transform of any dimension that supports this through attributes `units` and, optionally, `calendar`.
3. Set `dimnames()` in `hyper_array()`.

There has also been a significant simplification in `hyper_tibble()` because the `dimnames()` on the hyper_array are used to build the columns for the grid attributes.

One item of some concern: on `hyper_array()` the argument `drop = TRUE` as default causes some pain in the case where there is a degenerate dimension, such as in "ifremer/20171002.nc". It's easy enough to call the function with `drop = FALSE` but I think that should just be the default. If the data producers entered a dimension, then this package should not drop it by default. In the case of the "ifremer" file, the degenerate dimension contains the date of the observation so the information is actually useful. If so desired, the data user can always drop the dimension themselves.

This patch has been tested on CMIP5, CMIP6, CORDEX, ERA5, MODIS and assorted other NetCDF files, all successful. The results on the 7 NetCDF files included with the package were mixed:

- The "argo" files have their time information as character strings. I know there is an issue with that but I do not understand the issue so I can't fix it.
- "oceandata/S2008001.L3b_DAY_RRS.nc" massively failed with a `meta$error`, i.e. before any of the patch code is triggered. The file also won't open with `ncdf4::nc_open()` so the file appears to be corrupt.
- Other 4 files opened without problem.

I'm happy to discuss and fix the "argo" issue before going for a PR, unless you prefer otherwise - please advise.

My `roxygen` version is ahead of the package requirement (`RoxygenNote: 7.2.2`) so my installation refuses to rebuild documentation. As a consequence, I edited `NAMESPACE` manually.

**Comment by mdsumner** [comment]
oh thanks, this is awesome, very happy for you to fix the argo issue - I'll try to get a chance to look at examples but I'm confident in your assessment and I can patch up other problems as they come. 


don't worry about the S2008001.L3b_DAY_RRS.nc file that isn't supposed to work, but because netcdf and ncdf4 and RNetCDF contstantly changed in complicated ways it's in there as a way of safely erroring as a file type we can't actually read, but I haven't reviewed any of that in ages. 

 - L3b can't work because it's a complex netcdf with groups (I cant' remember the terminology, "complex types" maybe, but I always used an explicitly-HDF5 package for those data, see here it's in a test for no-vars, no-dims:  

https://github.com/ropensci/tidync/blob/main/tests/testthat/test-files-sources.R#L9

**Comment by pvanlaake** [comment]
Ok, NC_CHAR issue solved, including fixing an unnoticed error: Splitting NC_CHAR data should only be done when some but not all of the variables on the grid are NC_CHAR, in order to maintain consistent dimensions with variables having other data types. In cases where all variables on the grid are NC_CHAR, assume that these are all strings whose fixed length is recorded as the first dimension (that's just how NetCDF works). So no splitting, just drop the first dimension from the dimnames and all's well. You can easily check this by examining the argo files that have both types of grid.

Also circumvented the `drop = TRUE` issue by leaving degenerate dimensions upon reading and only dropping any dimensions after applying dimnames. I still think that `drop = FALSE` should be used as a matter of principle, but I can live with the current arrangement.

Going for a PR now.

**Comment by mdsumner** [comment]
I think I agree with you on the drop thing, forgot to say so  - I'll have a look at where it would bite me. 

The other thing I realized is that time-handling really belongs in ncmeta, but I'm more than happy to fold it in here and then over time I can worry about the refactor.

**Comment by mdsumner** [comment]
nice work!

![image](https://github.com/ropensci/tidync/assets/4107631/081136e4-1572-43cf-aa56-85bc2ee9af56)


I'll do a few checks, especially on downstream deps and probably just merge this in - thanks!

**Comment by pvanlaake** [comment]
> I think I agree with you on the drop thing, forgot to say so - I'll have a look at where it would bite me.
> 
> The other thing I realized is that time-handling really belongs in ncmeta, but I'm more than happy to fold it in here and then over time I can worry about the refactor.

I'll dig into `ncmeta` and see what can be done.

**Comment by pvanlaake** [comment]
I looked at `ncmeta` and, yes it can be done, but no `ncmeta` is not set up for it. As to the latter, in `ncmeta` the dimensions are not materialised, only the metadata is loaded. So in the code below there is nothing to attach the `CFtime` instance or the timestamps to:

```r
nc_meta.NetCDF <- function(x, ...) {
  (...)
  
  atts <- nc_atts(x) ## safe to assume there are always some attributes - it's the raison d'etre of NetCDF
  
  ## Add timestamp for any "time" dimension. Since not all files have a 
  ## "calendar" attribute or "axis == "T"", just try to create a CFtime
  ## instance for any dimension variable with a "units" attribute and a 
  ## "calendar", if present, and attach it to the appropriate dimension if 
  ## successful.
  idx <- which(atts$name == "units")
  vars <- atts$variable[idx]
  units <- unlist(atts$value[idx])
  for (i in seq_along(idx)) {
    if (nchar(units)[1] < 8) next
    cal <- unlist(atts$value[which(atts$variable == vars[i] & atts$name == "calendar")])
    try({
      cft <- CFtime::CFtime(units[i], cal)
      ## we have a CFtime instance so now read actual offsets and add
      offsets <- as.vector(var.get.nc(x, vars[i]))
      cft <- cft + offsets
      ## attach timestamp column to the appropriate dimension
      ?????????
    }, silent = TRUE)
  }

  structure(list(dimension = dims, 
       variable = vars, 
       attribute = atts, 
       axis = axis,
       grid = nc_grids_dimvar(dims, vars, axis)),
       class = "ncmeta")
}
```

Doing time handling in `ncmeta` thus needs a pretty significant design reorientation. The obvious choice would be to populate all dimensions with their values and could easily be implemented with existing `tidync` code but that is a design decision that you should carefully weigh. On the one hand I would say that it would be useful, but dependencies might break or have requirements beyond what a default implementation would offer.

**Comment by mdsumner** [comment]
I don't see the need to populate the dimensoin values, can't it just store the information? I'm not pushing back on you to do it, though. Honestly I'm not sure I'll have the motivation to pursue any of this anymore, I pretty much use GDAL for everything to do with NetCDF now (unless it's only 1D or compound types etc).  So, happy to take your advice on any things to change. Thanks for the feedback.

**Comment by pvanlaake** [comment]
I would suggest then to leave the `tidync` code as is, with the `CFtime` functionality as already merged in to the main branch. Close this issue as resolved?

**Comment by mdsumner** [comment]
yes for sure, will follow up 🙏
